daemon/k8s/init.go

```diff
@@ -137,8 +137,19 @@ func WaitForNodeInformation(ctx context.Context, log logrus.FieldLogger, localNo
 			logfields.K8sNodeIP:        k8sNodeIP,
 		}).Info("Received own node information from API server")
 
-		// If the host does not have an IPv6 address, return an error
-		if option.Config.EnableIPv6 && nodeIP6 == nil {
+		// error out in a dual-stack scenario without an IPv6 address set
+		//
+		// This is done to ensure that all nodes in a dual-stack cluster actually
+		// have IPv6 connectivity. The lack of this behaviour caused inconsistent
+		// fatal errors during certain migrations from single-stack IPv4 to dual-stack.
+		//
+		// We intentionally don't error out if only IPv6 is enabled since broken
+		// IPv6 connectivity would likely inhibit communication with the kube-apiserver
+		// in such a scenario. Instead we assume that the address is only unset
+		// temporarily and attempt to proceed anyways.
```

Comment:
I'm not sure I understand how this is safe. The call below to `useNodeCIDR(n)` is only ever called during startup as part of either this path or via `newDaemon` -> `configureIPAM`. As a result, the node's `IPv6AllocCIDR` will become nil in such environments. I don't see any reconciler to make this properly update later on.

As a general rule, it's a good idea to prevent surprising situations like that where there are uninitialized core fields somewhere in an internal structure when there's some issue during initialization. I don't see why in an IPv6 single-stack enviroment the CNI can't just wait until the cloud provider associates the relevant IP. Am I missing something?

Reply:
Hmm, you seem to be correct. This behaviour seems to match the current single-stack IPv4 behaviour though.

To answer your question on why we can't just wait for the cloud provider to set the relevant IP: in our environment there was a dependency of the vsphere-cpi on the CNI to be working; I don't exactly recall why that was the case though.

Reply:
There's a decent chance that we just never encountered an issue with single-stack IPv4 because the IPv4 address is always specified in practice. If this is the right place to do such checks, then I would expect IPv4-only case to do the equivalent check. That doesn't motivate why it's safe for IPv6.